handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -853,24 +921,38 @@ public final SSLEngineResult unwrap(
             int bytesProduced = 0;
             int bytesConsumed = 0;
             try {
-                for (; srcsOffset < srcsEndOffset; ++srcsOffset) {
+                srcLoop:
+                for (;;) {
                     ByteBuffer src = srcs[srcsOffset];
                     int remaining = src.remaining();
+                    final ByteBuf bioWriteCopyBuf;
+                    int pendingEncryptedBytes;
                     if (remaining == 0) {
-                        // We must skip empty buffers as BIO_write will return 0 if asked to write something
-                        // with length 0.
-                        continue;
+                        if (sslPending <= 0) {
+                            // We must skip empty buffers as BIO_write will return 0 if asked to write something
+                            // with length 0.
+                            if (++srcsOffset >= srcsEndOffset) {
+                                break;
+                            }
+                            continue;
+                        } else {
+                            bioWriteCopyBuf = null;
+                            pendingEncryptedBytes = SSL.bioLengthByteBuffer(networkBIO);
+                        }
+                    } else {
+                        // Write more encrypted data into the BIO. Ensure we only read one packet at a time as
+                        // stated in the SSLEngine javadocs.
+                        pendingEncryptedBytes = min(packetLength, remaining);
+                        bioWriteCopyBuf = writeEncryptedData(src, pendingEncryptedBytes);
                     }
-                    // Write more encrypted data into the BIO. Ensure we only read one packet at a time as
-                    // stated in the SSLEngine javadocs.
-                    int pendingEncryptedBytes = min(packetLength, remaining);
-                    ByteBuf bioWriteCopyBuf = writeEncryptedData(src, pendingEncryptedBytes);
                     try {
-                        readLoop:
-                        for (; dstsOffset < dstsEndOffset; ++dstsOffset) {
+                        for (;;) {
                             ByteBuffer dst = dsts[dstsOffset];
                             if (!dst.hasRemaining()) {
                                 // No space left in the destination buffer, skip it.
+                                if (++dstsOffset >= dstsEndOffset) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 5 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS134)